diff --git a/src/petablox/analyses/alloc/DomH.java b/src/petablox/analyses/alloc/DomH.java
index 8e05932..b9fbd8b 100755
--- a/src/petablox/analyses/alloc/DomH.java
+++ b/src/petablox/analyses/alloc/DomH.java
@@ -172,7 +172,11 @@ public class DomH extends ProgramDom<Object> {
             		break;
             	}
             }
-            int line = ((LineNumberTag)u.getTag("LineNumberTag")).getLineNumber();
+            int line = -1; // ((LineNumberTag)u.getTag("LineNumberTag")).getLineNumber();
+            Tag tg = u.getTag("LineNumberTag");
+            if(tg != null && tg instanceof LineNumberTag) {
+            	line = ((LineNumberTag) tg).getLineNumber();
+            }
             int mIdx = domM.indexOf(m);
             return "file=\"" + file + "\" " + "line=\"" + line + "\" " +
             "Mid=\"M" + mIdx + "\"" + " type=\"" + type + "\"";
diff --git a/src/petablox/util/ProcessExecutor.java b/src/petablox/util/ProcessExecutor.java
index 3632dc7..6486057 100755
--- a/src/petablox/util/ProcessExecutor.java
+++ b/src/petablox/util/ProcessExecutor.java
@@ -241,7 +241,9 @@ public final class ProcessExecutor {
                 }
                 os.flush();
             } catch (IOException ex) {
-                throw new RuntimeException(ex);
+				if(ex.toString().indexOf("Stream closed") == -1){
+                	throw new RuntimeException(ex);
+				}
             }
             
         }
